File: Algorithms/Selection_Algorithms/Quick_Select/QuickSelect.cpp
/**
* Finding Kth largest element in an array using QuickSelect algorithm
*/

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
       return findKthLargest(nums, 0, nums.size()-1, nums.size()-k);
    }

    int findKthLargest(vector<int>& nums, int left, int right ,int k) {
        int pivot = nums[right];
        int leftMark = left;
        for (int i = left; i<right; i++) {
            if(nums[i] <= pivot)
                swap(nums, leftMark++, i);
        }
        swap(nums, leftMark, right);

    if (leftMark == k)// Found kth smallest number
		return nums[leftMark];
	else if (leftMark < k)// Check right part
		return findKthLargest(nums, leftMark + 1, right, k);
	else // Check left part
		return findKthLargest(nums, left, leftMark - 1, k);
}

void swap( vector<int>& A, int i, int j) {
	int tmp = A[i];
	A[i] = A[j];
	A[j] = tmp;
}
};

File: Algorithms/Selection_Algorithms/Quick_Select/quick_select.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def find_kth_largest(nums, k)
    n = nums.size
    quick_select(nums, 0, n - 1, n - k) # Kth largest is at n - 1 - (k - 1) = (n - k)th index
end

def quick_select(a, p, r, k)            # Note the k here refers to (n - k) from the question stand-point
    return a[p] if p == r               # Base case of recursion
    q = rand_partition(a, p, r)         # Pick the pivot using random partition

    if k < q
        quick_select(a, p, q - 1, k)    # kth smallest is in the lower partition
    elsif k > q
        quick_select(a, q + 1, r, k)    # kth smallest is in the higher partition
    else
        a[q]                            # if q == k return Pivot as the kth smallest
    end
end

def partition(a, p, r)
    i = p - 1                           # Choose the index before the pth index
    pivot = a[r]                        # Choose the last element as pivot

    p.upto(r - 1) do |j|                # Loop from p to r-1 moving elements to either side of pivot
      next if a[j] > pivot              # Move to next element if element is already > than pivot element
      i += 1                            # Increment index i for current placement index
      a[i], a[j] = a[j], a[i]           # Swap elements at pivot placement index with the current index j
    end
  
    a[i + 1], a[r] = a[r], a[i + 1]     # Move the pivot to its final place: index before which all elements < pivot
    i + 1                               # return pivot index
end

def rand_partition(a, p, r)
    i = rand(p..r)                      # Pick a random index between p and r
    a[i], a[r] = a[r], a[i]             # Swap the last element with the one at i

    partition(a, p, r)                  # Return result of partition()
end
